src/EVM/Solidity.hs

```diff
@@ -358,28 +358,37 @@ readSolc pt root fp = do
   let contractName = T.pack $ takeBaseName fp
   case readJSON pt contractName fileContents of
       Left err -> pure . Left $ "unable to parse " <> show pt <> " project JSON: " <> fp
-        <> " Contract: " <> show contractName <> "\nError: " <> err
+        <> " Contract: " <> show contractName <> "\nError: " <> explain err
       Right (contracts, asts, sources) -> do
         conf <- readConfig
         when (conf.debug) $ liftIO $ putStrLn $ "Parsed contract: " <> show contractName <> " file: " <> fp
         sourceCache <- liftIO $ makeSourceCache root sources asts
         pure (Right (BuildOutput contracts sourceCache))
-
-yul :: Text -> Text -> IO (Maybe ByteString)
+  where
+    explain :: BytecodeReadingError -> String
+    explain (MissingJsonField field) = case field of
+      ast | ast == "ast" || ast == "sources" -> "missing field " <> (T.unpack ast) <> ". Recompile with `forge clean && forge build --ast"
+      _ -> "missing " <> (T.unpack field) <> " field"
+    explain (MissingOrInvalidJsonField field) = "missing or invalid " <> (T.unpack field) <> " field"
+    explain (InvalidSourceMap) = "invalid sourceMap format"
+    explain (UnlinkedLibrary contract) = "Unlinked libraries detected in bytecode, in " <> (T.unpack contract)
+    explain (OtherError contract errMsg) = "Error when reading bytecode of " <> (T.unpack contract) <> ": " <> T.unpack errMsg
```

Comment:
could this be extracted (maybe to a Show instance?) so it's usable by other tools?

Reply:
@elopez , I have added `Show` instance. Can you check if that is sufficient for you?